backend/src/main/java/gov/cdc/usds/simplereport/db/repository/TestEventRepository.java

```diff
@@ -20,18 +22,25 @@
 
 	public TestEvent findFirst1ByPatientOrderByCreatedAtDesc(Person p);
 
+    @EntityGraph(attributePaths = { "patient", "order" })
 	public TestEvent findByOrganizationAndInternalId(Organization o, UUID id);
 
 	// Need to control how this query is built. "between" is too vague.
 	// This is across all Orgs/facilities because datahub uploader users
 	@Query("FROM #{#entityName} q WHERE q.createdAt > :before AND q.createdAt <= :after ORDER BY q.createdAt DESC ")
 	public List<TestEvent> queryMatchAllBetweenDates(Date before, Date after);
 
-	@Query( value = "SELECT DISTINCT ON (test_order_id) * " +
-					" FROM {h-schema}test_event te " +
-					" WHERE te.facility_id = :facilityId " +
-					" AND te.created_at > :newerThanDate" +
-					" ORDER BY te.test_order_id, te.created_at desc", nativeQuery = true)
+    @Query(value = "WITH FILTEREDEVENTS AS (" +
+            " SELECT DISTINCT ON (test_order_id) * " +
+            " FROM {h-schema}test_event te " +
+            " WHERE te.facility_id = :facilityId " +
+            " ORDER BY test_order_id, te.created_at desc" +
+            ") " +
+            " SELECT * FROM FILTEREDEVENTS " +
+            " WHERE created_at > :newerThanDate " + // technically this could be in the CTE but then when we make it
+                                                    // more complicated somebody will break it
+            " ORDER BY created_at DESC ",
+            nativeQuery = true)
```

Comment:
So, this query will start with a fairly large select result for FILTEREDEVENTS, then narrow it down to a timerange (right now ~48hr || forever)? Seems like flipping it would be more efficient? 

However, where we really want to end up is with pagination and a different query to show just tests for a given patient. (This is the past test results in the patient page.)

Reply:
I suspect strongly that the optimizer does the same thing with either query, actually—I should take a look and see. My fear was that this query would end up used as the basis for something a little more complicated (such as filtering or pagination) and then bad things would happen. I'm happy to take a look at the query plan for this in production, though—could be fun :-D.

Reply:
I am wrong! The optimizer does nothing fancy here, so it does exactly what you expected. The difference in plan cost is small, but the difference in actual query cost is significantly higher (a representative query takes 5.8 ms instead of 1.3 ms to run). I'm not sure how long it would take for those milliseconds to add up to something we care about, but they definitely could.

```sql
explain
WITH FILTEREDEVENTS AS (
  SELECT DISTINCT ON (test_order_id) *  
  FROM test_event te  
  WHERE te.facility_id = $REDACTED  
  AND created_at > '2021-02-07'. 
  ORDER BY test_order_id, te.created_at desc
)                            
SELECT * FROM FILTEREDEVENTS
ORDER BY created_at DESC;

                                                               QUERY PLAN                                                                
-----------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=1553.09..1553.17 rows=34 width=304)
   Sort Key: filteredevents.created_at DESC
   CTE filteredevents
     ->  Unique  (cost=1551.37..1551.54 rows=34 width=1680)
           ->  Sort  (cost=1551.37..1551.46 rows=34 width=1680)
                 Sort Key: te.test_order_id, te.created_at DESC
                 ->  Seq Scan on test_event te  (cost=0.00..1550.51 rows=34 width=1680)
                       Filter: ((created_at > '2021-02-07 00:00:00'::timestamp without time zone) AND (facility_id = gen_random_uuid()))
   ->  CTE Scan on filteredevents  (cost=0.00..0.68 rows=34 width=304)
(9 rows)
```

Versus

```sql
explain
WITH FILTEREDEVENTS AS (
  SELECT DISTINCT ON (test_order_id) *  
  FROM test_event te    
  WHERE te.facility_id = gen_random_uuid()  
  ORDER BY test_order_id, te.created_at desc
)  
SELECT * FROM FILTEREDEVENTS
WHERE created_at > '2021-02-07'
ORDER BY created_at DESC;
                                       QUERY PLAN                                        
-----------------------------------------------------------------------------------------
 Sort  (cost=1571.04..1571.34 rows=121 width=304)
   Sort Key: filteredevents.created_at DESC
   CTE filteredevents
     ->  Unique  (cost=1556.87..1558.68 rows=363 width=1680)
           ->  Sort  (cost=1556.87..1557.78 rows=363 width=1680)
                 Sort Key: te.test_order_id, te.created_at DESC
                 ->  Seq Scan on test_event te  (cost=0.00..1541.43 rows=363 width=1680)
                       Filter: (facility_id = gen_random_uuid())
   ->  CTE Scan on filteredevents  (cost=0.00..8.17 rows=121 width=304)
         Filter: (created_at > '2021-02-07 00:00:00'::timestamp without time zone)
(10 rows)
```

The key, as you suspected, is the number of rows you can eliminate before you do the first (expensive) sort/group operation—if you eliminate nearly all of them, everything else goes much faster!